\documentclass[]{article}
\usepackage{amsmath} 

\begin{document}

\title{Site-Heterogeneous Character Change Models for Morphology}
\author{Wright, Pett, Students, Heath}
\date{Today}
\maketitle

\section{Introduction}
\subsection{Bayesian Modeling of Morphology}
	Recent interest in using Bayesian methods to model morphological evolution to estimate phylogenetic trees has spurred many studies into how well the existing toolkit performs, particularly in relation to parsimony methods.
	However, there has been little work expanding that toolkit. 
	The most common way to perform likelihood or Bayesian phylogenetic analysis using morphological data is by applying a model of morphological evolution called the Mk model.
	Many workers using morphological data have raised questions about the realism of this model.
	In this paper, we will discuss methodological advancements aimed at improving the realism of the model. \par
	The Mk model was introduced by Lewis in 2001.
	As a generalization of the Jukes-Cantor model, it makes the same set of assumptions: that change between any two states is equally likely, that the stationary character frequencies of every character are the same, and that each character is always in one of \textit{k} states.
	The model also makes the Markovian assumption that a character can change instantaneously, regardless of previous states. 
	Any one of these assumptions may strike a reader as problematic for their particular dataset. \par
	Heterogeneity in the evolutionary process is difficult to accommodate in morphological data. 
	Making concrete \textit{a priori} statements about the relative probabilities of transitions between character states in a morphological character matrix is challenging. 
	Unlike a matrix of nucleotide characters, a `0' character at one state may mean that the trait never evolved in one lineage, but was lost in another.
	This lack of common meaning complicates the interpretation of how change occurs over time.
	Likewise, the magnitude of change from one state to another may be different between characters, even if the characters mean the same thing (i.e. `0' representing loss and `1' representing gain).
	For example, in one character, changing from a `0' state to a `1' state could represent a change at a single locus, but at another could represent a much larger change coordinated across many genes.
	This has complicated the ability to specify $\alpha$ parameters to the Q-matrix \textit{a priori}.
	This is a stark contrast to molecular data, in which base pairs or amino acid residues are generally assumed to have similar properties across an alignment. \par
Prior work extending the Mk model has focused on relaxing the assumption of equal character frequencies at stationarity.
MrBayes implemented a parameter called the Symmetric Dirichlet Hyperprior, which allowed users to place a prior or hyperprior on state frequencies. 
The probability of observing a change in a character is dependent not only on the probability of change from this character to another, but on the frequency of the starting character.
Even if a character has a low probability of changing, change in that character may still be observed many times if the stationary frequency of that character is high (i.e., the character is observed many times). 
Likewise, a highly-probable change may be seen relatively rarely if the starting character is observed rarely. 
Therefore, relaxing the assumption of equal stationary frequencies has been a way of changing the probabilities of observing different changes without making strong \textit{a priori} statements about transition probabilities. \par
In the case of binary data, the prior used was a Beta prior (the Dirichlet is a generalization of the Beta distribution), which operated in a fashion similar to gamma-distributed rate variation: the distribution is discretized into a user-specified number of categories, the median forward and backward transition rate are calculated for each category, and the likelihood of the character is calculated over each category and summed to form the total likelihood. 
For binary data, the state frequencies are integrated out of the likelihood, making the symmetric Dirichlet a prior.
For multistate data, the state frequencies are drawn from a Dirichlet distribution.
The parameter was referred to as a hyperprior because users can place a distribution on the parameter to the Beta distribution, and because the state frequencies are not integrated out of the likelihood function for multistate data. 
As implemented in MrBayes, the Beta distribution was assumed to be symmetric, meaning that if there were characters for which the stationary frequency of 0 is high, there would also be corresponding characters for which the stationary frequency of 1 is high.\par
This is a very useful model extension for morphological evolution.
There are clear contexts to improve this concept by borrowing from the molecular literature.
The CAT model of Lartillot represents one way forward. 
This model, implemented in PhyloBayes, uses a Dirichlet process to assign individual sites to categories, which differ in their stationary frequencies. 
In this model, the number of such categories that describe the data is a free parameter. 
The use of a Dirichlet prior allows for flexibility in the number of states, as opposed to a Beta prior which is limited to binary data.
However, many datasets that have been analyzed under the CAT model are phylogenomic datasets of thousands of loci, as compared to small morphology datasets.\par
In RevBayes, we have implemented a number of useful extensions to the Dirichlet prior of MrBayes, and a constrained version of the CAT model. 
Our extension to the symmetric Beta prior allows for the Beta distribution to be asymmetrical.
This removes the need for there to be equal numbers of characters with high stationary frequencies for a character state as there are characters with low stationary frequencies for that same character state.
Our CAT-like model (hereafter Site-Heterogeneous Discrete Morphology (SDHM) model) is a finite mixture model that allows characters to fit into a user-defined number of character categories that differ in their stationary frequencies. 
The bins draw their stationary frequencies from a Dirichlet to accommodate multistate data. 

\subsection{Morphological Phylogeny of the Formicidae}

To test the efficacy of our analytical techniques, we use an ant dataset created from datasets from Barden and Grimaldi and one from Keller.
Ants are ecologically crucial organisms as both interacting partners for a variety of plants and animals, and as shapers of the ecosystem via soil cycling and nest building.
As such, they have attracted much work in the world of molecular systematics.
Ants have also have a rich fossil record, with most subfamilies being represented. 
This fossil record has been used for systematic work, as well as a variety of other ecological and evolutionary questions.\par
The monophyly of major subfamilies has been consistently supported by most molecular and morphological studies, with the exception of the Cerapachyinae. 
Molecular systematics had originally shaken up the ant tree of life considerably, breaking up clades considered to be monophyletic based on morphological work.
For example, based on morphological work, six current subfamilies had been previously considered to be a single subfamily, the Ponerinae, until molecular evidence indicated that two of the clades (Ectatomminae and Heteroponerinae) within that subfamily were demonstrated to be more closely related to other ants on the tree.
The Ponerinae was then broken into six distinct subfamilies, one of which is called Ponerinae. 
Recent molecular work has continued to support these six subfamilies as monophyletic, though their relationships to one another remain poorly supported. \par
Excellent morphological matrices on the ingroup of the Formicidae have been available since the early '90s. 
Recent morphological matrices have expanded the sampling of ants from in the large Ponerine family, which was previously undersampled.
Sampling of fossil ants was also expanded to include specimens from the Cretaceous.
These ants were concluded to be stem lineages, and some did not demonstrate any particular taxonomic affinity. 
Even with expanded taxon sampling in the Ponerinae, there is still conflict between the molecular and morphological estimates of phylogenetic relationships. \par
In this study, we combine the extant matrices of Keller, and the extinct-extant matrix of Barden and Grimaldi to expand the taxon and character sampling.  
This dataset has interesting properties.
Because there are stem ant lineages represented, there are taxa that have characteristics that are lost after the divergence of the stem lineages.
This extremely one-sided loss structure violates the Mk model assumption of equal transition rates. 
We would expect for these characters to be better modeled by a model that can accommodate asymmetrical transition rates.
Some apomorphies of the ant group are also gained after the divergence of the stem lineages, which also violates the assumption of equal change probabilities.
Because of these model violations, this dataset is an excellent test case for models that relax assumptions of the Mk model. \par
We use Bayes Factor model selection to assess the fit of these relaxed models to the data.
Using this dataset, we strongly support that the use of models that relax key assumptions of the Mk model can greatly improve the fit of the model to the data.
We also perform simulations to demonstrate that the true number of transition rate asymmetry categories is detectable from the data.

\section{Methods}

\subsection{Modeling site-heterogeneous state frequencies}

\subsubsection{Discretized beta model}
We model site-specific state frequencies by assuming the state frequency $\pi_i$ at site $i$ is drawn from a Beta prior distribution with hyperparameters $\alpha,\beta$.
Then, the likelihood of the data $D$ is computed by marginalizing over the value for $\pi_i$ at each site.

\begin{equation}\label{continuous-likelihood}
f(D\mid \alpha,\beta) = \prod_{i=1}^n\int_0^1 f(D_i \mid \pi_i)f(\pi_i \mid \alpha,\beta)d\pi_i
\end{equation}

To simplify the computation, we approximate this integrated likelihood by assuming $\pi_i$ is drawn from a mixture distribution, where the mixture categories are defined deterministically by discretizing a Beta distribution into $k$ bins whose boundaries are defined using $k-1$ quantiles.
The the value of mixture category $j$ is indicated by $\beta_j$ and is computed as the interquantile median of the $j$-th bin.
In other words, $\beta_j = I^{-1}_{(2j-1)/2k}(\alpha,\beta)$, where $I_x(\alpha,\beta)$ is the regularized incomplete Beta function.
The mixing proportion is $1/k$ for each category.
Then the likelihood is computed by summing over the $k$ discrete mixture categories for $\pi_i$ at each site.

\begin{equation}\label{discretized-likelihood}
f(D\mid \alpha,\beta) = \prod_{i=1}^n\frac{1}{k}\sum_{j=1}^k f(D_i \mid \pi_i = \beta_j)
\end{equation}

Note that in the limit as the number of mixture categories approaches infinity, equation \ref{discretized-likelihood} is identical to equation \ref{continuous-likelihood}.

\subsubsection{Site-heterogeneous discrete multistate model}

When analyzing multistate character data with $S$ character states, we assume the state frequencies $\pi_i$ at each site $i$ are drawn from a Dirichlet distribution with concentration parameters $\alpha_1,\ldots,\alpha_S$.
Again, we use a mixture distribution to simplify the computation, but instead of defining the mixture categories using a discretized Dirichlet distribution, the value $\delta_j$ of mixture category $j$ is assumed to be drawn from the same underlying Dirichlet distribution with parameters $\alpha_1,\ldots,\alpha_S$.
The mixing proportions are $\theta_j$, which are themselves drawn from a uniform Dirichlet prior distribution.

$$f(D\mid \delta_1,\ldots,\delta_k,\theta_1,\ldots,\theta_k) = \prod_{i=1}^n \sum_{j=1}^k  f(D_i \mid \pi_i = \delta_j)\theta_j$$

\subsection{Data Matrices}
\subsubsection{Empirical Matrices}
Several large and well-documented ant matrices were used in this study.
The first was that of Keller (2011). 
This matrix is of extant ant groups. 
Keller's matrix was collected with special attention to the poneromorph subgroups (Amblyoponinae, Ectatomminae,Heteroponerinae, Paraponerinae, Ponerinae, and Proceratiinae) and has 139 characters and 105 taxa.
The matrix is highly complete for a morphological dataset, with about 60\% of cells being filled.
Of the total character set, 100 characters were binary and 39 characters were multistate.
Because of the scope of Keller's study, the taxon sampling is biased towards the poneromorphs, and away from the other ant subfamilies, including large subfamilies such as the Dolichoderinae and the Formicinae. \par
We also used data from Barden and Grimaldi (2015). 
This matrix contained both extant and extinct ants, and was collected in order to place stem ants from the Cretaceous period.
One crown ant amber fossil, \textit{Kyromyrma neffi}, was included in this matrix, but the key feature of this matrix is the sampling of the stem group, including multiple samples from the \textit{Gerontoformica} genus.
This matrix also expands sampling in the non-ponerine groups that are underrepresented in the Keller matrix.
The Barden-Grimaldi matrix contained 42 characters and 41 taxa. 
Of these characters, 26 were binary.\par
In order to achieve maximum coverage, we merged these character matrices.
The taxonomic overlap between the two was 11 taxa, and all subfamilies of the Formicidae except the extinct Formiciinae were represented in the matrix. 
Eleven characters were represented in both matrices without any recoding, and both matrices had many characters that were inapplicable to the other matrices.
Inapplicable characters were left in the matrix due to their ability to resolve bipartitions in the the authors' respective groups of interest. \par
Six characters were recoded to make the character states uniform between the two matrices.
Most of these changes simply involved changing the terminology used in reference to the character states.
For example, Barden and Grimaldi's character 15 is the same as Keller's character 47, but the two matrices had inverse character codings relative to one another.
The remainder of the changes were changing binary characters to a multistate in cases where one author had more states than the other. 
We will refer to this dataset as the combined dataset. \par
To test the SHDM model, we used several permutations of these data.
First, to test the binary version of the model, we used Barden and Grimaldi's dataset with the multistate characters removed.
This dataset was chosen because the dataset size and completeness is very typical to a morpholgoical dataset, particularly one involving fossils.
We used this to test the SHDM model.
We used the combined matrix to also test the full SHDM model.
Lastly, we used the combined matrix, stripped of multi-state characters to test the binary model on a larger dataset.
The two datasets including both multistate and binary characters were used to examine the effect of dataset size on the ability to estimate model parameters. \par

\subsubsection{Simulated Matrices}

To test the performance of the SHDM model on idealized datasets, we used RevBayes to simulate five sets of 100 replicates of morphological data.
The first two sets of simulations were based on Barden and Grimaldi's dataset size and tree.
Under this set of conditions, datasets of 41 taxa and 42 characters were generated in RevBayes.
In one set, only binary characters were evolved under the Beta model.
In the other, both binary and mutistate characters were evolved.
For both datasets, the binary or Dirichlet distribution was discretized into four categories.
We also created one set of 100 replicates that was generated under a strict Mk model with no state frequency variation to see if the model will overfit state frequency variation where none exists.\par
The other set of two simulations was based on the full datasets.
One set of 139 binary characters was generated for the full 105 taxa under the binary model; a second set including multistate characters was generated under the SHDM model.
The tree used to generate the data was the tree estimated from the combined empirical dataset using the SHDM. \par

\subsection{Phylogenetic Analyses}
\subsubsection{Empirical Data}

Trees were estimated from each empirical dataset using multiple models of evolution.
The Mk model was used for each dataset, with an appropriately-sized Q-matrix.
For the binary-only datasets, the binary model was used.
The datasets containing both binary and multistate characters were analyzed using the SHDM, with the dimensionality of the Dirichlet process being equal to the largest number of character states. 
Each dataset was run under several different numbers of discrete state frequency variation categories, from two to eight categories.
All datasets were corrected for not observing invariant characters. \par
For the binary datasets, Bayes Factor model fitting was used to assess what the optimal number of discrete categories were.
We used stepping-stone model fitting to calculate the marginal likelihood.
We first ran an MCMC analysis to determine how many generations are required to attain convergence (about one million).
Each of the 50 stepping stones was then run for one million generations.
Due to computational limits, for the multistate datasets, we used the mean of the MCMC distribution to compare across different discretizations of the model. \par


\subsubsection{Simulated Data}

Because the true tree cannot be known from empirical data, we used simulated data to observe the effect this model has on accuracy of estimation, and to see if it is possible to detect the true number of discrete categories of state frequency variation.
To this end, for each replicate, all datasets were estimated in RevBayes using the true model.
Trees were also estimated from all datasets without allowing any state frequency variation (i.e., the Mk model) to simulate the effect of excluding variation (underparameterizing the data) completely.
Lastly, every dataset was run under the correct model (either the binary model or the SHDM model), but with a misspecified number of categories.
The number of categories ran from two to eight.
These categories were chosen because they run from half the true number to double the true number. 
In total, each dataset was run under one true model and eight misspecified models. \par

\section{Results}

\subsection{Empirical Phylogenetic Analyses}
\subsection{Binary Data}

The marginal likelihood comparisons for the Barden and Grimaldi dataset, when multistate characters are removed can be seen in Fig. 1a.
The marginal likelihood comparisons for the combined dataset with the multistate characters removed can be seen in Fig. 1b.




	a) Best fit model?
	b) Tree comparisons, esp. branch lengths. No asc. bias correction + misspecified model, no correction + misspecified model, correction + correct model, correction + misspecified model 
	c) Tree set visualization comparisons
	
\section{Discussion}	

	a) How has the modeling of morphology changed since the Mk was proposed by Lewis? What aspects are comparable to how we model DNA? Which are not? What are potentially interesting future directions? 
	b) General discussion of usage of treeset visualization
	c) something else.

\section{Conclusion}	


\section{Ackno}	


\end{document}